[otbnsim] Assorted improvements

imphil opened the pull request:
A large set of small changes to improve otbnsim in various ways. These are random things I found when trying to generate more statistics from simulation runs. (The actual PR implementing that will come shortly after a bit more testing.)

Please have a look at the individual commit messages for more details.

imphil commented on hw/ip/otbn/dv/otbnsim/sim/insn.py:
```diff
@@ -218,7 +218,7 @@ def execute(self, state: OTBNState) -> None:
         val2 = state.gprs.get_reg(self.grs2).read_unsigned()
         if val1 == val2:
-            state.pc_next = self.offset
+            state.set_pc_next(self.offset)
```
Note: We have a couple options to implement this behavior. If our goal is to make the code use `state.pc_next` instead of an explicit setter we can use a property setter (this code shows up in the docs). But then we have other set_* functions in the OTBNState class currently, so we would loose a bit of consistency there. I'm fine either way.

rswarbrick replied:
I'm happy to use a setter, but can we call it `set_next_pc`? I think that might be a bit clearer in the documentation (where readers won't know anything about a variable called "pc_next")

rswarbrick commented on hw/ip/otbn/dv/otbnsim/sim/loop.py:
```diff
@@ -47,6 +47,18 @@ def __init__(self, start_addr: int, insn_count: int, restarts: int):
         self.start_addr = start_addr
         self.match_addr = start_addr + 4 * insn_count
 
+    @property
```
Is there a strong reason to use properties rather than e.g. `get_loop_insn_addr()`? We've not used them elsewhere and I think I probably prefer being really explicit about what's the state that we're storing and what's derived from it.

imphil replied:
I like properties as they make refactorings such as the ones I've been doing here easier (no need to adjust the callers). But I really don't care much, switched over to getters/setters.

rswarbrick commented on hw/ip/otbn/dv/otbnsim/sim/loop.py:
```diff
@@ -47,6 +47,18 @@ def __init__(self, start_addr: int, insn_count: int, restarts: int):
+        assert self.start_addr >= 4
+        return self.start_addr - 4
+
+    @property
+    def last_insn_addr_in_loop(self) -> int:
+        ''' The address of the last instruction in a loop body. '''
+        assert self.match_addr >= 4
+        return self.match_addr - 4
```
Since you're changing the `step` function to compare `pc` with the last instruction in the loop (an improvement, thanks!), maybe it would be better to store "`last_addr`" rather than `match_addr`, which would get rid of this computation entirely?

imphil replied:
Yes, that's clearly better. Changed to that. (Went around in circles there a couple times.)

rswarbrick commented on hw/ip/otbn/dv/otbnsim/sim/state.py:
```diff
@@ -58,25 +58,35 @@ def __init__(self) -> None:
         self._new_rnd_data = None  # type: Optional[int]
         self._urnd_reseed_complete = False
 
+    @property
+    def pc_next(self) -> int:
+        if self._pc_next_override:
+            return self._pc_next_override
+        return self.pc + 4
```
As in previous commits, please can we just make this an explicit function (`get_next_pc` maybe?)

rswarbrick commented on hw/ip/otbn/dv/otbnsim/sim/state.py:
```diff
@@ -58,25 +58,35 @@ def __init__(self) -> None:
         c = []  # type: List[Trace]
         c += self.gprs.changes()
-        if self.pc_next is not None:
+        if self._pc_next_override:
```
This needs an `is not None` (because the zero address is falsy in Python)

rswarbrick commented on hw/ip/otbn/dv/otbnsim/sim/state.py:
```diff
@@ -188,9 +198,6 @@ def check_jump_dest(self) -> None:
         If not, generates a BadAddrError.
 
         '''
-        if self.pc_next is None:
-            return
-
```
I think this changes how we spot code that runs off the top of memory. I *think* it's right, but maybe we can delete some check somewhere else?

imphil replied:
Really? The `self.get_next_pc() >= self.imem_size` check is still there. 

The `if self.pc_next is None` check limited this function (`check_jump_dest()`) to the cases where we were pc_next was explicitly set (as result of a jump, branch, or loop). After this change, the function runs always, even if pc_next is just pc + 4. Since that's now trivial to do I chose to expand the check as small safety net.

imphil replied:
> we must have been checking for straight-line code falling off the top of memory in some other way.

Ah, now I understand. I had a look and couldn't find the check, it looks like we didn't check for that before.

Good point about renaming, I did that.

rswarbrick replied:
Oops! Nice catch :-)

rswarbrick commented on hw/ip/otbn/dv/otbnsim/test/stats_test.py:
```diff
@@ -7,32 +7,35 @@
 
-    assert os.path.exists(asm_file)
-    elf_file = testutil.asm_and_link_one_file(asm_file, tmpdir)
+    sim = testutil.prepare_sim_for_asm_file(asm_file, tmpdir, start_addr = 0)
```
Nit: I think the usual style is `start_addr=0` for keyword arguments?

rswarbrick reviewed (commented):
Thanks for these tidy-ups! I've got a few nitty comments (and "hey, could we do even better?" questions), but this looks good.

imphil commented:
Thanks for your review @rswarbrick! 

I addressed your comments and additionally added a doc rewrite pass to transform the `state.set_next_pc()` call into the same syntax we had before (`PC <= offset`).

(And I found that `git absorb` (https://crates.io/crates/git-absorb) worked great!)

rswarbrick commented on hw/ip/otbn/dv/otbnsim/sim/stats.py:
```diff
@@ -37,33 +37,40 @@ def record_stall(self) -> None:
+        '''Record the execution of an instruction.
+
+        insn is the currently executed instruction. state_bc is the state of
+        OTBN the instruction is committed.
```
Nit: Missing "before", I think.

rswarbrick commented on hw/ip/otbn/dv/otbnsim/sim/stats.py:
```diff
@@ -37,33 +37,40 @@ def record_stall(self) -> None:
         # - Indirect function calls: jalr x1, <grs1>, 0
-        if (isinstance(insn, JAL) or isinstance(insn, JALR)) and insn.grd == 1:
-            call_stack = state_before_commit.peek_call_stack()
+        if is_jump and insn.grd == 1:  # type: ignore
```
That's sad. Presumably mypy can't follow the types through setting `is_jump`. Is there any reason to change this from what was there before, which didn't need the waiver?

imphil replied:
is_jump is going to be used again in the basic block stats PR (together with a is_branch) and I want to avoid duplicating this code. Unfortunately mypy got priority 2 behind "reducing code duplication".

I tried a couple options but I couldn't really find a readable workaround that wouldn't require effectively duplicating the JAL/JALR check.

rswarbrick replied:
Oh, I see. Boo :-(

rswarbrick reviewed (approved):
This is a definite improvement to the code, thanks! A couple of remaining comments / questions, but I'm happy for this to go in either way.
